pkg/errors/errors.go

```diff
@@ -35,27 +36,53 @@ func init() {
 	}
 }
 
+func MakeErrID() string {
+	errUUID, err := uuid.NewRandom()
+	if err == nil {
+		return errUUID.String()
+	}
+	return fmt.Sprintf("<error ID could not be constructed: %s>", err)
+}
+
 // Interceptor returns a grpc.UnaryServerInterceptor that inspects outgoing
 // errors and logs (to "logger") and replaces errors that are not *status.Status
 // instances or status instances that indicate an internal/unknown error.
 func Interceptor(logger *zap.Logger) grpc.UnaryServerInterceptor {
 	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
 		resp, err = handler(ctx, req)
-		statusErr, ok := status.FromError(err)
-
-		switch {
-		case !ok:
-			logger.Error("encountered non-Status error during unary server call", zap.String("method", info.FullMethod), zap.Error(err))
-			err = status.Error(codes.Internal, fmt.Sprintf("a non-Status error occurred: %s", err))
-		case statusErr.Code() == codes.Internal, statusErr.Code() == codes.Unknown:
-			logger.Error("encountered internal Status error during unary server call",
-				zap.String("method", info.FullMethod),
-				zap.Stringer("code", statusErr.Code()),
-				zap.String("message", statusErr.Message()),
-				zap.Any("details", statusErr.Details()),
-				zap.Error(err))
-			if statusErr.Code() == codes.Internal && obfuscateInternalErrors {
-				err = errInternal
+
+		if err != nil {
+			// Separate the root cause from the stacktrace wrapping.
+			trace := err.Error()
+			rootErr := stacktrace.RootCause(err)
+
+			statusErr, ok := status.FromError(rootErr)
+			switch {
+			case !ok:
+				errID := MakeErrID()
+				logger.Error(
+					fmt.Sprintf("encountered non-Status error %s during unary server call", errID),
```

Comment:
`logger.Errorf` should give you the same functionality.

Reply:
`type *zap.Logger has no field or method Errorf`; what did you have in mind?

Reply:
More specifically: https://godoc.org/go.uber.org/zap#Logger

Reply:
I was thinking about `zap.SugaredLogger` but it turns out that it does not support additional, unformatted fields. With that, comment resolved.